### PR TITLE
client: only when trying to clear suid/sgid will allow perm check pass

### PR DESCRIPTION
**Pull request by lxbsz**
When an unprivileged user tries to change the file mode, which the corresponding mode is already there, for example:

$ ll
-rw-r--r-- 1 root   root       0 Jun  1 12:36 fileA
$ su ${unprivileged-user} -c "chmod a+r fileA"

Currently it will succeed. While this is incorrect and won't comply with the Posix and the existing filesystems.

We should fail it anyway and only when it's trying to clear the suid/sgid will pass the permission check.

Introduced-by db10cfc8077("client: allow unprivileged users to clear suid/sgid")
Fixes: https://tracker.ceph.com/issues/61552
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -5900,7 +5900,7 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
 	 *
 	 * Only allow unprivileged users to clear S_ISUID and S_ISUID.
 	 */
-	(m & ~(S_ISUID | S_ISGID)))
+	((m & ~(S_ISUID | S_ISGID)) || !(m & (S_ISUID | S_ISGID))))
```
Comment (mchangir) [inline src/client/Client.cc:5903]:
```
if (perms.uid() != 0 && perms.uid() != in->uid &&
    ((m & ~(S_ISUID | S_ISGID)) || !(m & (S_ISUID | S_ISGID))))
```
My reading of the above condition is that even if the file is owned by root, a non-privileged user can unset the `S_ISUID` and `S_ISGID` bits. But is that POSIXly correct ?
Please correct me if my understanding of the above condition is wrong.
Here's what I got to see during testing of this PR ...
```
mchangir:build$ echo hello > /mnt/mycephfs/greeting
mchangir:build$ ls -l /mnt/mycephfs/greeting
-rw-r--r--. 1 mchangir mchangir 6 Sep 20 12:30 /mnt/mycephfs/greeting
mchangir:build$ sudo chown root:root /mnt/mycephfs/greeting
mchangir:build$ ls -l /mnt/mycephfs/greeting
-rw-r--r--. 1 root root 6 Sep 20 12:30 /mnt/mycephfs/greeting
mchangir:build$ sudo chmod u+s /mnt/mycephfs/greeting
mchangir:build$ ls -l /mnt/mycephfs/greeting
-rwSr--r--. 1 root root 6 Sep 20 12:30 /mnt/mycephfs/greeting
mchangir:build$ chmod u-s /mnt/mycephfs/greeting
mchangir:build$ ls -l /mnt/mycephfs/greeting
-rw-r--r--. 1 root root 6 Sep 20 12:30 /mnt/mycephfs/greeting
```

Such a transition isn't allowed by btrfs on my `/home` partition.

Comment (lxbsz) [reply src/client/Client.cc:5903]: @mchangir I think you found a corner case bug. Thanks very much.

Then we must pass the `ATTR_KILL_SUID/ATTR_KILL_SGID` from kernel space via the libfuse to fix this issues, except that I couldn't find a better way to fix this issue.

Comment (mchangir) [reply src/client/Client.cc:5903]: I could mostly find privileges dropping code in the kernel code during the `write` op.
Didn't go through it exhaustively.

You might want to add a small test case for this corner case.

Comment (lxbsz) [reply src/client/Client.cc:5903]: The xfstest-dev have already done that. So I think we can just use that instead of adding a new one.

**Pull request closed without merging**